### DISCUSSION TITLE
Equivalent of "ccache -c" not issued during runtime?

### DISCUSSION THREAD

**Discussion by zephyrus00jp** [ccache/ccache#177] [Support]
I have noticed an issue of ccache running out of the cache space during a compilation of Mozilla Thunderbird  despite the fact that I set max_size = 5.0 G.
In the following the parameter list from |ccache -p|
and the output of |ccache -s| in a typical condition. This was after a |ccache -c| was issued.

I have started to notice this cache overflow problem since the spring of 2016 (last year), I think.
The local build never experienced this before that time frame.

If I run |ccache -c| to clear the cache area, then the build will proceed, often to successful completion.
(Note that it is no -C but -c.)
So actually, now I inserted |ccache -c| at the start of a build command, whic initiates and build mozilla thunderbird,  AND at the end of the script, too.
Still I run out of the space from time to time.

Granted that mozilla thunderbird is such a big package and I am running |make -j4| so there *is* a chance of more than  instances of |make| creating a large output in the cache simultaneously, but
given that we are talking  about a few hundred MB of area restored by my manual |ccache -c| when the
overflow condition is noticed, and that C++ compilation command is issued separately for source files during that time (and so |ccache| is invoked multiple times)  I think something went awry regarding ccache's bhavior regarding |-c| flag.
Quote from |man ccache| :
       -c, --cleanup
           Clean up the cache by removing old cached files until the specified
           file number and cache size limits are not exceeded. This also
           recalculates the cache file count and size totals. Normally, there
           is no need to initiate cleanup manually as ccache keeps the cache
           below the specified limits at runtime and keeps statistics up to
           date on each compilation. Forcing a cleanup is mostly useful if you
           manually modify the cache contents or believe that the cache size
           statistics may be inaccurate.
End quote.

It says  that "Normally, there  is no need to initiate cleanup manually" and I believe it was the behavior up until around the early spring of 2016.

Is there a smoking gun or was there a related issue before?

The listing: 

ccache -p
(default) base_dir = 
(environment) cache_dir = /FF-NEW/cache-dir
(default) cache_dir_levels = 2
(default) compiler = 
(default) compiler_check = mtime
(default) compression = false
(default) compression_level = 6
(default) cpp_extension = 
(default) direct_mode = true
(default) disable = false
(default) extra_files_to_hash = 
(default) hard_link = false
(environment) hash_dir = true
(default) ignore_headers_in_manifest = 
(default) keep_comments_cpp = false
(default) limit_multiple = 0.8
(environment) log_file = /FF-NEW/cache-dir/logfile
(default) max_files = 0
(/FF-NEW/cache-dir/ccache.conf) max_size = 5.0G
(default) path = 
(default) prefix_command = 
(default) prefix_command_cpp = 
(default) read_only = false
(default) read_only_direct = false
(default) recache = false
(default) run_second_cpp = true
(default) sloppiness = 
(default) stats = true
(default) temporary_dir = 
(default) umask = 
(default) unify = false

ccache -s
cache directory                     /FF-NEW/cache-dir
primary config                      /FF-NEW/cache-dir/ccache.conf
secondary config      (readonly)    /etc/ccache.conf
cache hit (direct)                204080
cache hit (preprocessed)           96355
cache miss                        665297
cache hit rate                     31.11 %
called for link                    46770
called for preprocessing           36230
multiple source files                  4
compile failed                      4913
ccache internal error                191
preprocessor error                  9839
couldn't find the compiler             1
bad compiler arguments               591
unsupported source language          853
autoconf compile/link              21011
unsupported compiler option           45
no input file                     194153
cleanups performed                  5952
files in cache                     12147
cache size                           4.0 GB
max cache size                       5.0 GB

TIA

**Comment by jrosdahl** [comment]
> I have noticed an issue of ccache running out of the cache space during a compilation of Mozilla Thunderbird despite the fact that I set max_size = 5.0 G.

I'm not sure I understand. What do you mean with "running out of the cache space"? Does the build fail? If so, how?

> cleanups performed 5952

This counter is increased for each triggered cleanup (both explicitly using `ccache -s` and implicitly due to cache size becoming close enough to the limit), so it seems that cleanups are performed.

**Comment by zephyrus00jp** [comment]
I think I found the cause.
Somehow ccache tries to keep the used cache size to 4.0 GB and it would allow the growth to 5.0 GB.
Starting early Spring last year (2016), I think other files used up the file system where ccache's cache is stored, and thus there was LESS THAN 1GB remaining. It seems to me that before ccache's 5GB limit was reached, the file system overflowed. (I saw no AUTOMATIC cleaning message a la
[2017-06-02T00:57:07.115740 8003 ] Cleaning up cache directory /FF-NEW/cache-dir/a

Previously, all the cleaning-up message was for the directory, /0, /1, /2, ..., /f in this order in the close proximity of time. This suggests they were all from explicit "ccache -c" invocation.

So I reduced the max size to 4GB and now I have more than 1GB available in the file system where the cache is stored. 
Maybe because of the reduced size of the max size (??), I now see a clean up message of directory in radom order. E.g. 

[2017-06-02T00:54:24.549599 6316 ] Cleaning up cache directory /FF-NEW/cache-dir/d
[2017-06-02T00:55:34.010278 7056 ] Cleaning up cache directory /FF-NEW/cache-dir/b
[2017-06-02T00:57:07.115740 8003 ] Cleaning up cache directory /FF-NEW/cache-dir/a
[2017-06-02T00:57:26.411545 8200 ] Cleaning up cache directory /FF-NEW/cache-dir/8
[2017-06-02T00:57:39.496612 8233 ] Cleaning up cache directory /FF-NEW/cache-dir/3
[2017-06-02T00:59:43.607542 8771 ] Cleaning up cache directory /FF-NEW/cache-dir/0
[2017-06-02T01:00:29.188994 9026 ] Cleaning up cache directory /FF-NEW/cache-dir/6
[2017-06-02T01:01:50.858348 9322 ] Cleaning up cache directory /FF-NEW/cache-dir/5
[2017-06-02T01:02:05.112047 9438 ] Cleaning up cache directory /FF-NEW/cache-dir/9
[2017-06-02T01:02:27.562960 9460 ] Cleaning up cache directory /FF-NEW/cache-dir/e

I think we ought to catch this type of setting that may overflow due to the short remaining size (this can happen due to the unexpected growth of other files in the file system over the years) and print a warning or something to that effect. The condition to trigger such a warning is
The remaining space in the file system where the cache is stored < The max cache size - the current size
We may even want to throw in a fuzz factor of 100MB or something like that to trigger the warning early.

Just a thought. This warning could have saved me a lot of grief over the 12 months when I found the large job aborting in midway due to the file system overflew. Yes, admittedly, it is a user configuration issue, but it was not entirely clear to me what was incorrect over these months until I got enough of it and sat down and wrote a bug report and analyzed the issue :-)

I can create a patch if it seems acceptable.

TIA

**Comment by afbjorklund** [comment]
@zephyrus00jp: So, ccache never used more than 5.0G but _something else_ filled up the disk ?

Adding file system monitors is not very portable (or easy), so probably better done elsewhere.
However, ccache should never create half-a-file or something but continue to "work" (fallback)

You could also keep the ccache directory on a dedicated partition, to help with the bookkeeping ?

**Comment by jrosdahl** [comment]
I don't think that it ccache should warn about the file system becoming full for two reasons:

1. What makes ccache different in this regard? By the same reasoning, all programs that write to the disk should have similar warnings. That just doesn't make sense.
2. ccache deliberately (almost) never prints warning messages (i.e., messages to stderr but with exit code 0) since such messages can confuse e.g. configure scripts that try to probe whether the compiler works. Messages on stderr from the compiler have been known to trip up the configure script from autotools, for example.
